samples/99_zil_interpreter/app/builtins.rb

```diff
@@ -7,15 +8,51 @@ def define_for_evaled_arguments(&implementation)
   }
 end
 
+# helper method to check for argument count
+# to be used with rethrowing the error with the `ROUTINE` name in front
+# I could do it with metaprogramming, but I doubt anyone would want to see that XD
+def argc!(arguments, count, cmp_type = :==)
+  case cmp_type
+  when :===
+    raise ArgumentError, "`count` (#{count}) not a `Range` object" unless count.is_a?(Range)
+    raise FunctionError, "has an arity (#{count.begin}≤n≤#{count.end}), got #{arguments.length}" unless count === arguments.length
+  else
+    raise FunctionError, "has an arity (n#{cmp_type}#{count}), got #{arguments.length}" unless arguments.length.send(cmp_type, count)
+  end
+end
+
+def expect_argument_count!(arguments, expected_count)
+  argc!(arguments, expected_count)
+end
+
+# this is painful... either I make it work, and make `argc!` unintuitive or rewrite `argc!` to do a check ;-;
+# guess I'll rewrite it
+# rewritten, it looks not-too-bad
+def expect_argument_count_in_range!(arguments, range)
+  argc!(arguments, range, :===)
+end
+
+def expect_minimum_argument_count!(arguments, min_count)
+  argc!(arguments, min_count, :>)
```

Comment:
Shouldn't that be `:>=` ?

Reply:
that is correct :facepalm:

Reply:
I actually don't know how I managed to get it wrong